[llvm-lit][test] Precommit tests for curly braces in lit internal shell

connieyzhu opened the pull request:
This patch creates tests to check lit's failure to execute curly braces {}. This is a prequisite patch to https://github.com/llvm/llvm-project/pull/102830.

connieyzhu committed:
[llvm-lit][test] Precommit tests for curly braces in lit internal shell

This patch creates tests to check lit's failure to execute curly braces
{}.

connieyzhu commented:
CC: @ilovepi @petrhosek @jh7370 

Hi there, I'm writing some tests for curly braces in lit's internal shell before I implement the functionality in https://github.com/llvm/llvm-project/pull/102830. I wasn't sure if I should write more in-depth tests in this patch, since the same syntax error will be given for every test at the moment, or if I should just include more tests in my later patch once I have implemented the proper functionality.

ilovepi commented:
There's some value in showing that each test used to fail, but now passes. I'm not sure it needs to be particularly exhaustive, but whatever basic tests you're considering are probably reasonable to include in this patch. Typically its nice if none of the RUN lines change, but you can easily spot a difference in the test behavior. I know I appreciate that in the middle or backend tests I review.

connieyzhu committed:
[llvm-lit][test] Added tests for syntax errors and nested braces

ilovepi commented on llvm/utils/lit/tests/Inputs/shtest-shell/curly-brace.txt:
```diff
@@ -0,0 +1,20 @@
+# TWO-CMDS: bar
+
+## Test nested curly brace.
+# RUN: { echo foo; { echo bar; echo baz; }; } | FileCheck --check-prefix=NESTED %s
+# RUN: { { echo foo; echo bar; }; echo baz; } | FileCheck --check-prefix=NESTED %s
```
I'm not sure this is checking things well enough. The nested `{...}` probably needs to result in something that's distinct from the regular case. As written, you cant distinguish the nested version from the un-nested version. Maybe you can use a `|` or file redirection + `cat`?

ilovepi commented on llvm/utils/lit/tests/shtest-shell.py:
```diff
@@ -4,7 +4,7 @@
 # buildbots.
 # RUN: cat %t.out
-# RUN: FileCheck --input-file %t.out %s
+# RUN: FileCheck --input-file %t.out %s --dump-input-context=1000
```
leftover from debugging?

connieyzhu replied:
Yes, forgot to delete that. I will remove it in my next change.

connieyzhu committed:
[llvm-lit][test] Updated nested curly brace test

connieyzhu committed:
[llvm-lit][test] Moved fail tests into individual files

This patch moves tests that check for errors into separate files so that
the each error can be accounted for.

ilovepi commented on llvm/utils/lit/tests/Inputs/shtest-shell/curly-brace.txt:
```diff
@@ -0,0 +1,19 @@
+# TWO-CMDS: bar
+
+## Test nested curly brace.
+# RUN: { echo foo > %t/temp.txt; { echo bar; echo baz; }; cat %t/temp.txt; } | FileCheck --check-prefix=NESTED %s
```
I don't see a `mkdir %t`, so I'm not sure this works as intended. If you don't need other files, you can just use `%t`.

ilovepi commented on llvm/utils/lit/tests/Inputs/shtest-shell/curly-brace.txt:
```diff
@@ -0,0 +1,19 @@
+## Test nested curly brace.
+# RUN: { echo foo > %t/temp.txt; { echo bar; echo baz; }; cat %t/temp.txt; } | FileCheck --check-prefix=NESTED %s
+
+# NESTED: bar
+# NESTED: baz
+# NESTED: foo
```
```suggestion
#      NESTED: bar
# NESTED-NEXT: baz
# NESTED-NEXT: foo
```

connieyzhu committed:
[llvm-lit][test][NFC] Formatting changes

jh7370 commented:
I'm not opposed to this PR being developed, but I don't want it and the follow-on PR landed without the RFC discussion I've requested in the source issue: as noted there, I'm not convinced this functionality is needed in lit (not going to elaborate further here, to avoid splitting the discussion).
